Parallel compact block downloads, take 3

instagibbs opened the pull request:
This is an attempt at mitigating https://github.com/bitcoin/bitcoin/issues/25258 , which is a revival of https://github.com/bitcoin/bitcoin/pull/10984, which is a revival of https://github.com/bitcoin/bitcoin/pull/9447.

This PR attempts to mitigate a single case, where high bandwidth peers can bail us out of a flakey
peer not completing blocks for us. We allow up to 2 additional getblocktxns requests per unique block.
This would hopefully allow the chance for an honest high bandwidth peer to hand us the transactions
even if the first in flight peer stalls out.

In contrast to previous effort:

1) it will not help if subsequent peers send block headers only, so only high-bandwidth peers this time. See: https://github.com/bitcoin/bitcoin/pull/10984/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R1411
2)  `MAX_GETBLOCKTXN_TXN_AFTER_FIRST_IN_FLIGHT` is removed, in favor of aiding recovery during turbulent mempools
3) We require one of the 3 block fetching slots to be an outbound peer. This can be the original offering peer, or subsequent compact blocks given by high bandwidth peers.

instagibbs committed:
alias BlockDownloadMap for mapBlocksInFlight

instagibbs committed:
Remove nBlocksInFlight

instagibbs committed:
Convert mapBlocksInFlight to a multimap

instagibbs committed:
Only request full blocks from the peer we thought had the block in-flight

This is a change in behavior so that if for some reason we request a block from a peer, we don't allow an unsolicited CMPCT_BLOCK announcement for that same block to cause a request for a full block from the uninvited peer (as some type of request is already outstanding from the original peer)

ajtowns commented on src/net_processing.cpp:
```diff
@@ -4452,13 +4465,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                     expected_blocktxn = true;
                     break;
                 }
+                first_in_flight = false;
```
This logic is different to the previous `first_in_flight`; probably should be the same?

instagibbs replied:
done, cleaner thank you

ajtowns reviewed (commented):
Seems okay at first glance. It would be nice to have a "blockrequest.cpp" module that collects all this logic together. Perhaps we want to keep this change simple so it's a smaller backport, and refactor later, though.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4300,7 +4300,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-            if (fAlreadyInFlight) {
+            if (in_flight_same_peer) {
                 // We requested this block for some reason, but our mempool will probably be useless
                 // so we just grab the block via normal getdata
```
sounds right, I'll sit on this for a bit

ajtowns replied:
Probably better to defer it as an unrelated change anyway?

dergoegge commented on src/net_processing.cpp:
```diff
@@ -4364,8 +4364,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-                    if (req.indexes.size() <= MAX_GETBLOCKTXN_TXN_AFTER_FIRST_IN_FLIGHT || first_in_flight) {
+                    // as long as it's first, or not too large, and as long as we chose
+                    // the peer for high-bandwidth relay.
+                    auto hb_peer = std::find(lNodesAnnouncingHeaderAndIDs.begin(), lNodesAnnouncingHeaderAndIDs.end(), pfrom.GetId());
+                    if ((hb_peer != lNodesAnnouncingHeaderAndIDs.end() && req.indexes.size() <= MAX_GETBLOCKTXN_TXN_AFTER_FIRST_IN_FLIGHT) ||
```
Could use `pfrom.m_bip152_highbandwidth_to`?

instagibbs replied:
done

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4275,12 +4283,21 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
+        auto range_flight = mapBlocksInFlight.equal_range(pindex->GetBlockHash());
+        bool fAlreadyInFlight = range_flight.first != range_flight.second;
+        bool in_flight_same_peer = false;
```
don't love your name suggestion; changed the initialization though

ajtowns replied:
Let me rephrase. What I meant was that `in_flight_same_peer` confused me: "we just received the block from this peer, of course it was in flight, how else would it get to us?". But what the bool is actually telling us is that we had deliberately asked this peer for this block, it wasn't just spontaneously sent to us.

instagibbs replied:
ok, I can see that and how it reads cleaner, changed

ajtowns commented on src/net_processing.h:
```diff
@@ -22,6 +22,11 @@ static const bool DEFAULT_PEERBLOOMFILTERS = false;
+/** Greg: Setting to 3 to match number of high bandwidth peers... ~ensures one is outbound */
+static const unsigned int MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK = 3;
+/** Maximum number of missing transactions for a non-first getblocktxn request */
+static const unsigned int MAX_GETBLOCKTXN_TXN_AFTER_FIRST_IN_FLIGHT = 10;
```
I think the limit's fine; when you start up with an empty mempool, you'll be requesting most of the transactions in the block because your mempool's empty, and there's no need to do that three times in parallel. Could perhaps have the limit work the other way: if you've already got N txs and need K txs, and `0 < K <= N/4 + 10`; then if the block just has ten 100kvB txs, you won't request them multiple times.

dergoegge replied:
Imo, it doesn't seem worthwhile to have the limit:
* the good case (small K) should be far more common than the bad one (large K)
* the limit makes relay of blocks with large K less reliable w.r.t to stalling (reduces to what we have now)
* the limit adds code complexity and a maintenance burden. 10 is not a good value according to your data and if #10984 had been merged then we would need to re-evaluate now. Who's to say this won't change again.

instagibbs replied:
if the issue is during warmup, that's a pretty small window, and typically won't actually do parallel fetching, 3 seems rare even when warmed up with limit of 10. I'll gather some more stats on that front, seeing how many parallel blocks are initialized.

instagibbs replied:
This "warmup" should really only be an issue after we've seen new blocks from 3+ peers since we don't persist hb peers across restarts and we have to wait for `MaybeSetPeerAsAnnouncingHeaderAndIDs` to be called on multiple unique peers. I now suspect the warmup period handles itself? My logs show 4+ blocks per node with only one partial block.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -901,7 +901,7 @@ class PeerManagerImpl final : public PeerManager
     void FindNextBlocksToDownload(const Peer& peer, unsigned int count, std::vector<const CBlockIndex*>& vBlocks, NodeId& nodeStaller) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-    typedef std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator>> BlockDownloadMap;
+    typedef std::multimap<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator>> BlockDownloadMap;
```
after attempting the change, I think that would violate https://github.com/bitcoin/bitcoin/pull/27626/commits/e424dcba60e61f014b19a3e2f59238ca3b91f3db#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R1394 which requires insertion ordering to know which is the stalling peer.

totally forgot about the FIXME...

dergoegge commented on src/net_processing.cpp:
```diff
@@ -4342,8 +4372,17 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+                    // We will try to round-trip any compact blocks we get on failure,
+                    // as long as it's first, or not too large and as long as we chose
+                    // the peer for high-bandwidth relay.
+                    if (first_in_flight ||
+                        (pfrom.m_bip152_highbandwidth_to && req.indexes.size() <= MAX_GETBLOCKTXN_TXN_AFTER_FIRST_IN_FLIGHT)) {
+                        req.blockhash = pindex->GetBlockHash();
+                        m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::GETBLOCKTXN, req));
```
This could be a `else if` of the outer if statement. That would avoid the nesting here.

instagibbs replied:
done

instagibbs commented:
I'm clean up history later when we're closer to ACKs

Sjors commented:
> In contrast to previous effort, it will not help if subsequent peers send block headers only

Can you elaborate on this? Are you saying that falling back to full block download is pointless?

Sjors commented:
So that earlier attempt would solicit a compact block from a (BIP 152) low-bandwidth mode peer, in response to it announcing a header. That's similar to what we already do in `HeadersDirectFetchBlocks`, but allowing more than 1 in transit?

Whereas in this PR we focus on the high-bandwidth mode peers, which are [allowed to send us](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#intended-protocol-flow) a `cmpctblock` message straight away (rather than waiting for us to ask). Right now we can only have one `cmpctblock` -> `getblocktxn` -> `blocktxn` "session" in progress. This PR makes that 3.

This approach seems fine, but the other attempts were never merged, which is why I am confused by the "it will not help" phrase. Are you saying these earlier attempts failed in some experiment, are they useless in general or is this new approach just better (and less complex?)?

instagibbs commented:
> This PR makes that 3.

Makes it 3 parallel block downloads, first can be of any kind, the 2nd and 3rd must be compact block announcements from high badwidth peers. 

> I am confused by the "it will not help" phrase

This PR does not take advantage of low-bandwidth compact block peers message patterns, in other words.

instagibbs commented:
I was getting some sybil-stalls at <=10 limit, so for now uncapped it.

I also added in logic to ensure that at least one of the 3 slots is taken by an outbound peer. Added more extensive tests.

instagibbs committed:
Support up to 3 parallel compact block txn fetchings

A single outbound slot is required, so if the first two slots
are taken by inbound in-flights, the node will reject additional
unless they are coming from outbound.

This means in the case where a fast sybil peer is attempting to
stall out a node, a single high bandwidth outbound peer can
mitigate the attack.

instagibbs committed:
Add tests for parallel compact block downloads

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -1132,8 +1147,8 @@ void PeerManagerImpl::RemoveBlockRequest(const uint256& hash, std::optional<Node
 
-    // Currently we don't request more than one peer for same block
-    Assume(mapBlocksInFlight.count(hash) == 1);
+    // We should never requested too many of this block
```
nit: perhaps "We should not have requested too many of this block"

instagibbs replied:
done

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -1179,8 +1194,8 @@ bool PeerManagerImpl::BlockRequested(NodeId nodeid, const CBlockIndex& block, st
-    // Make sure it's not listed somewhere already.
-    RemoveBlockRequest(hash, std::nullopt);
+    // Make sure it's not being fetched already from same peer.
+    RemoveBlockRequest(hash, nodeid);
```
Is it worth adding an `Assume()` here that we have strictly less than MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK requests outstanding?

instagibbs replied:
done

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -4460,33 +4496,43 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                 return;
             }
 
+            bool first_in_flight = already_in_flight == 0 || (already_in_flight == 1 && requested_block_from_this_peer);
```
So if we get 3 compact block announcements from 3 peers, and we then send 3 getblocktxn messages to them, and then the first blocktxn comes back and reconstruction fails, we will wait until we get the third blocktxn message back before we send out a getdata message for the full block, I think?

That seems suboptimal, though I don't have a fix in mind yet.

sdaftuar replied:
Actually, can we use the insertion order guarantees of multimap to solve this?

instagibbs replied:
Looks like some lost in translation logic from prior PRs. Indeed, we can just peek at the first entry, if it exists. Fixed, I think.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -4277,15 +4307,25 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            }
+            range_flight.first++;
+        }
+        bool first_in_flight = already_in_flight == 0 || (already_in_flight == 1 && requested_block_from_this_peer);
```
This notion of `first_in_flight` is not correct, I think.  Imagine we get a headers announcement from an outbound (non-hb) peer, and so we send a getdata(cmpctblock).  In the meantime, we get a compact block announcement from an inbound peer which fails to reconstruct.  Then the compact block comes back from the outbound peer; `first_in_flight` will be false because we think we have two requests outstanding.

Further down, when we decide whether to send a GETBLOCKTXN message to this non-HB outbound peer, we will choose not to, because `first_in_flight` is false and the peer is not already HB.  This seems like it would be a regression, as inbound peers would gain the ability to interfere with compact block relay happening with other peers just by making a compactblock announcement.

sdaftuar replied:
Similarly here, can we use the insertion order of multimap to determine which peer was actually first to make the announcement?  If in the situation where an outbound (non-hb) peer is the first to announce that we still always get the block, then this new logic would not be making anything worse, which seems sufficient to me.

instagibbs replied:
> If in the situation where an outbound (non-hb) peer is the first to announce that we still always send the getblocktxn if compact blocks fail, then this new logic would not be making anything worse, which seems sufficient to me.

Exactly. Fixed, I think!

sdaftuar reviewed (commented):
I have not reviewed the test or done any manual testing yet, but just finished reading the code.

One conceptual thing that I wanted to mention: right now it is possible for none of our HB compactblock peers to be outbound. The way the HB outbound reservation logic works is that if we ever have an outbound peer taking an HB slot, we will not allow inbounds to take over all 3 slots. However, it is possible that a peer starting up for the first time could have all its slots taken by inbounds.

instagibbs commented:
> However, it is possible that a peer starting up for the first time could have all its slots taken by inbounds.

Yeah, goes without saying, but maybe it doesn't. This PR offers no protection on fresh restart(no hb peers), and only offers 3 parallel downloads when an outbound is selected as hb and offers a block

instagibbs commented on src/net_processing.cpp:
```diff
@@ -1125,34 +1126,40 @@ bool PeerManagerImpl::IsBlockRequested(const uint256& hash)
-        // Block was not requested
+    auto range = mapBlocksInFlight.equal_range(hash);
+    if (range.first == range.second) {
+        // Block was not requested by any peer
```
done

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4320,7 +4320,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         // If we're not close to tip yet, give up and let parallel block fetch work its magic
-        if (!fAlreadyInFlight && !CanDirectFetch()) {
+        if (!in_flight_same_peer && !CanDirectFetch()) {
```
hmm yeah, this is old code, doesn't make sense to me either. removed the change.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4460,33 +4496,43 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+                    invs.push_back(CInv(MSG_BLOCK | GetFetchFlags(*peer), resp.blockhash));
+                    m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::GETDATA, invs));
+                } else {
+                    RemoveBlockRequest(resp.blockhash, pfrom.GetId()); // Reset in-flight state in case of whitelist
```
removed nonsense copy/paste

sdaftuar commented:
Thanks for the quick updates -- the code looks right to me now; will test.

sdaftuar commented:
ACK 2d12fab42cc0f11d5ede96c7eb71a496931df0b5. I've reviewed the test as well and have been running this branch live, and it looks good to me.

I think it'd be good to merge this sooner than later so we can get some experience running this in master before it appears in a release, so that we can get a sense of whether we want to institute any better bandwidth management before deploying.

fanquake commented:
> "jq: command not found" weird one, probably just restart?

Rebase should do it. Added recently for IWYU in the tidy job.

instagibbs commented:
rebased

sdaftuar commented:
ACK 42c2696ae58e07de005edf1dda952761f8c9008e

brunoerg commented on test/functional/p2p_compactblocks.py:
```diff
@@ -823,12 +827,100 @@ def assert_highbandwidth_states(node, hb_to, hb_from):
+            return block, cmpct_block
+
+        # First, make delivery_peer, inbound_peer, and outbound_peer high bandwidth
+        block, cmpct_block = announce_cmpct_block(node, delivery_peer, 1)
```
Suggestion to simplify it, in case you touch it again:
```diff
index 50cce9318..cd9268f6b 100755
--- a/test/functional/p2p_compactblocks.py
+++ b/test/functional/p2p_compactblocks.py
@@ -847,29 +847,14 @@ class CompactBlocksTest(BitcoinTestFramework):
             return block, cmpct_block
 
         # First, make delivery_peer, inbound_peer, and outbound_peer high 
bandwidth
-        block, cmpct_block = announce_cmpct_block(node, delivery_peer, 1)
-        msg = msg_blocktxn()
-        msg.block_transactions.blockhash = block.sha256
-        msg.block_transactions.transactions = block.vtx[1:]
-        delivery_peer.send_and_ping(msg)
-        assert_equal(int(node.getbestblockhash(), 16), block.sha256)
-        delivery_peer.clear_getblocktxn()
-
-        block, cmpct_block = announce_cmpct_block(node, inbound_peer, 1)
-        msg = msg_blocktxn()
-        msg.block_transactions.blockhash = block.sha256
-        msg.block_transactions.transactions = block.vtx[1:]
-        inbound_peer.send_and_ping(msg)
-        assert_equal(int(node.getbestblockhash(), 16), block.sha256)
-        inbound_peer.clear_getblocktxn()
-
-        block, cmpct_block = announce_cmpct_block(node, outbound_peer, 1)
-        msg = msg_blocktxn()
-        msg.block_transactions.blockhash = block.sha256
-        msg.block_transactions.transactions = block.vtx[1:]
-        outbound_peer.send_and_ping(msg)
-        assert_equal(int(node.getbestblockhash(), 16), block.sha256)
-        outbound_peer.clear_getblocktxn()
+        for peer in [delivery_peer, inbound_peer, outbound_peer]:
+            block, _ = announce_cmpct_block(node, peer, 1)
+            msg = msg_blocktxn()
+            msg.block_transactions.blockhash = block.sha256
+            msg.block_transactions.transactions = block.vtx[1:]
+            peer.send_and_ping(msg)
+            assert_equal(int(node.getbestblockhash(), 16), block.sha256)
+            peer.clear_getblocktxn()
 
         # Test the simple parallel download case...
         for num_missing in [1, 5, 20]:
```

instagibbs replied:
done

dergoegge reviewed (approved):
Code review ACK 42c2696ae58e07de005edf1dda952761f8c9008e

Haven't done a lot of testing yet besides reviewing the tests and running a node with this patch, but the code looks good to me.

(comments are nits, feel free to ignore)

instagibbs commented:
holding off on touching for nits, can do in follow-up

ajtowns commented on src/net_processing.cpp:
```diff
@@ -4338,9 +4386,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                     req.blockhash = pindex->GetBlockHash();
                     m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::GETBLOCKTXN, req));
+                } else if (pfrom.m_bip152_highbandwidth_to &&
+                    (!nodestate->m_is_inbound ||
```
Should be `!pfrom.IsInboundConn()` presumably, to include manual connections at least.

instagibbs replied:
took @ajtowns suggestion

sdaftuar replied:
This doesn't result in a behavior change, does it?  As far as I can tell `nodestate->m_is_inbound` is initialized to the value of `pfrom.IsInboundConn()` when the `CNodeState()` is constructed -- is that understanding correct?

I'm good with avoiding the use of CNodeState where possible though, if that is the reason for this change.

If these are different somehow, then perhaps we'd want to change the logic in `IsBlockRequestedFromOutbound` to match it.

dergoegge replied:
> This doesn't result in a behavior change, does it? As far as I can tell nodestate->m_is_inbound is initialized to the value of pfrom.IsInboundConn() when the CNodeState() is constructed -- is that understanding correct?

Yes that is correct, it does not change behavior (my original suggestion would have changed it slightly, as it would exclude manual conns but that wasn't intentional).

> I'm good with avoiding the use of CNodeState where possible though, if that is the reason for this change.

That was the only reason for this nit from my side.

mzumsande reviewed (commented):
Concept ACK, code looks good on first walkthrough.

I want to run this code with some additional logging for a day or two before acking.

In particular, I wonder if the fact that we'll now try additional peers will lead to redundancies / increased traffic under non-adversarial conditions where there's no staller but delivering the block just needs some time - especially when blocks contain a large number of txns that weren't in our mempool, or on slower networks like Tor. Did you consider the possibility of giving the original peer a few seconds to deliver before requesting from additional peers?

Also, all that `first->second.second` in connection with the multimap is not very pretty or easy to read, but I don't see a better way.

ajtowns commented:
> In particular, I wonder if the fact that we'll now try additional peers will lead to redundancies / increased traffic under non-adversarial conditions where there's no staller but delivering the block just needs some time - especially when blocks contain a large number of txns that weren't in our mempool, or on slower networks like Tor. Did you consider the possibility of giving the original peer a few seconds to deliver before requesting from additional peers?

Yeah, this is the behaviour I'm seeing --

```
# peer=3 first reports it
2023-05-22T03:13:48.490251Z Saw new header hash=000000000000000000028a6e01ad99705343eea1ff85e43aa580bbae52a3a8ab height=790827
2023-05-22T03:13:48.490389Z [net] Saw new cmpctblock header hash=000000000000000000028a6e01ad99705343eea1ff85e43aa580bbae52a3a8ab peer=3
2023-05-22T03:13:48.512908Z [cmpctblock] Initialized PartiallyDownloadedBlock for block 000000000000000000028a6e01ad99705343eea1ff85e43aa580bbae52a3a8ab using a cmpctblock of size 18890
2023-05-22T03:13:48.513282Z [net] sending getblocktxn (35 bytes) peer=3

# peer=11 0.2s later
2023-05-22T03:13:48.740077Z [net] received: cmpctblock (18890 bytes) peer=11
2023-05-22T03:13:48.761214Z [cmpctblock] Initialized PartiallyDownloadedBlock for block 000000000000000000028a6e01ad99705343eea1ff85e43aa580bbae52a3a8ab using a cmpctblock of size 18890
2023-05-22T03:13:48.761562Z [net] sending getblocktxn (35 bytes) peer=11

# peer=9 another 0.2s later
2023-05-22T03:13:48.946673Z [net] received: cmpctblock (18890 bytes) peer=9
2023-05-22T03:13:48.968735Z [cmpctblock] Initialized PartiallyDownloadedBlock for block 000000000000000000028a6e01ad99705343eea1ff85e43aa580bbae52a3a8ab using a cmpctblock of size 18890
2023-05-22T03:13:48.969097Z [net] sending getblocktxn (35 bytes) peer=9

# peer=11 responds first
2023-05-22T03:13:49.051028Z [net] received: blocktxn (97486 bytes) peer=11
2023-05-22T03:13:49.055872Z [cmpctblock]   blocktxn peer=11 tx=f401366ed3a40be59923388aa5e58d90c921ac162a92d14a573d24553ba0ce65
2023-05-22T03:13:49.055891Z [cmpctblock]   blocktxn peer=11 tx=5ebf80ab6ac479bd3568cf9d1575c62b94aaf8bace4058970ca5e3e35a3b0ab1
2023-05-22T03:13:49.081909Z [cmpctblock] Successfully reconstructed block 000000000000000000028a6e01ad99705343eea1ff85e43aa580bbae52a3a8ab with 1 txn prefilled, 3078 txn from mempool (incl at least 0 from extra pool) and 2 txn requested

# then peer=3 about 0.4s later
2023-05-22T03:13:49.457226Z [net] received: blocktxn (97486 bytes) peer=3
2023-05-22T03:13:49.461365Z [net] Peer 3 sent us block transactions for block we weren't expecting

# then peer=9 about 0.14s later
2023-05-22T03:13:49.600309Z [net] received: blocktxn (97486 bytes) peer=9
2023-05-22T03:13:49.604364Z [net] Peer 9 sent us block transactions for block we weren't expecting
```

For block 790788 I received two 1.5MB blocktxn messages:

```
2023-05-21T20:02:38.285186Z Saw new header hash=000000000000000000023a845351079f13015d10aa55003d30e7589e597caa2a height=790788
2023-05-21T20:02:38.285264Z [net] Saw new cmpctblock header hash=000000000000000000023a845351079f13015d10aa55003d30e7589e597caa2a peer=3

# peer=9 announces 0.3s later
2023-05-21T20:02:38.659630Z [net] received: cmpctblock (4129 bytes) peer=9
2023-05-21T20:02:38.679677Z [cmpctblock] Initialized PartiallyDownloadedBlock for block 000000000000000000023a845351079f13015d10aa55003d30e7589e597caa2a using a cmpctblock of size 4129

# peer=3 took ~2s to respond
2023-05-21T20:02:40.240719Z [net] received: blocktxn (1541605 bytes) peer=3
2023-05-21T20:02:43.332966Z UpdateTip: new best=000000000000000000023a845351079f13015d10aa55003d30e7589e597caa2a height=790788 version=0x20014000 log2_work=94.193595 tx=841476965 date='2023-05-21T20:01:57Z' progress=1.000000 cache=26.9MiB(224222txo)

# peer=9 took ~7s to respond
2023-05-21T20:02:45.830208Z [net] received: blocktxn (1541605 bytes) peer=9
2023-05-21T20:02:45.893063Z [net] Peer 9 sent us block transactions for block we weren't expecting
```

In that case, if peer=9 had announced first, and I'd waited 5s before requesting for peer=3, I'd still have doubled up the 1.5MB response, and not gotten either respond until 7s after the announcement. But at least there would have been a maybe 50% chance of avoiding the second request entirely.

Tripling the outgoing bandwidth load for peers selected for hb compact block relay seems concerning; not sure if I'd call it a blocker though.

instagibbs commented:
@ajtowns This is a delicate trade-off as you're aware.

> delivering the block just needs some time 

There's a lot of mix and match we can likely do to improve the bandwidth usage, probably with only a small delay in block propagation. There are block timeouts, yes, but we also care about block prop speed for mining fairness(e.g., selfish mining attacks). So maybe there can be 2 buckets for each type of issue.

e.g.,
1) first or second slot must have outbound peer (best effort to keep blocks fast)
2) third request is only dispatched after 5s delay (don't let node get unduly timed out)

Obviously a bit more complexity, but the linked issue can stay open and discussion continue with metrics?

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4338,9 +4386,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+                    // ... or it's a hb relay peer and:
+                    // - peer is outbound, or
+                    // - we already have an outbound attempt in flight(so we'll take what we can get), or
+                    // - it's not the final parallel download slot (which we may reserve for first outbound)
```
https://github.com/bitcoin/bitcoin/pull/27626/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4346 <--- this line will filter out any attempts beyond 3.

So implicitly each of these conditions in this block hold with `already_in_flight < MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK`

I could touch up my comments, add some `Assume`s to make things clearer, if indeed the code is doing what I think it is.

ajtowns replied:
Well I was right in saying "I don't understand this logic" -- it's just about reserving the "final" slot for the outbound peer.

ajtowns replied:
I guess I think it might be better to not do multiple parallel requests to outbound peers (on the one hand hopefully your outbounds are trustworthy and will respond quickly; on the other hand it's undesirable if listening nodes end up having to send up to 3x as much data when non-listening nodes request blocktxns from all their hb peers), but otherwise this makes sense.

instagibbs replied:
added more clarifying comment and an `Assume` for now

instagibbs replied:
> I guess I think it might be better to not do multiple parallel requests to outbound peers

That makes some sense since using an outbound is mostly for the anti-sybil portion of the reasoning. Once we have an outbound hb peer, they're likely highly reliable, and we probably don't need to try so hard for other connections. Maybe even allow only up to two attempts if we already have an outbound in flight.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -1117,40 +1121,57 @@ std::chrono::microseconds PeerManagerImpl::NextInvToInbounds(std::chrono::micros
+    for (auto range = mapBlocksInFlight.equal_range(hash); range.first != range.second; range.first++) {
+        auto [nodeid, block_it] = range.first->second;
+        CNodeState *nodestate = State(nodeid);
+        assert(nodestate);
```
done

instagibbs commented on src/net_processing.cpp:
```diff
@@ -1117,40 +1121,57 @@ std::chrono::microseconds PeerManagerImpl::NextInvToInbounds(std::chrono::micros
-        // Last validated block on the queue was received.
-        m_peers_downloading_from--;
+        CNodeState *state = State(node_id);
+        assert(state != nullptr);
```
done

Sjors replied:
It ended up in 03423f8bd12b95a06a4a9d8377e781625dd38aae

instagibbs commented on test/functional/p2p_compactblocks.py:
```diff
@@ -823,12 +827,100 @@ def assert_highbandwidth_states(node, hb_to, hb_from):
+        for num_missing in [1, 5, 20]:
+
+            # Remaining low-bandwidth peer is stalling_peer, who announces first
+            [peer['bip152_hb_to'] for peer in node.getpeerinfo()] == [False, True, True, True]
```
fixed

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4456,18 +4496,23 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
-            bool expected_blocktxn = false;
             auto range_flight = mapBlocksInFlight.equal_range(resp.blockhash);
+            size_t already_in_flight = std::distance(range_flight.first, range_flight.second);
```
leaving as-is, reads more naturally to me

instagibbs commented:
responded to all nits

let's keep the issue open for further discussion on bandwidth saving strategies

ajtowns commented:
> > Tripling the outgoing bandwidth load for peers selected for hb compact block relay seems concerning; not sure if I'd call it a blocker though.
> 
> To be clear, this is only a tripling in the case where where all your peers are roughly equally as fast, and triple of _what_ is the most important thing.

In an "everyone's honest" environment, I think the most likely reason for there to be enough of a delay between `cmpctblock / getblocktxns / blocktxns` that there'll be another `cmpctblock` received in the meantime (and thus another `getblocktxns / blocktxns`) is that the `blocktxns` message is large due to a lot of the block being made up of non-mempool txs (violating standardness, or fees paid out of band). That's already pretty rare, and it's already size limited by max block weight. The difference between limiting the number of reqs to outbounds is exacerbated by the ratio of listening/non-listening -- if there's 80k non-listening nodes and 20k listening nodes, then the worst case has 240k requests being made to the hb listening nodes from non-listening nodes instead of just 80k (as a non-listening node doesn't have any inbound peers in the first place), and that extra load might not just be affecting random non-listening nodes.

I think this is more of a todo than a blocker though.

ajtowns reviewed (approved):
ACK 3b57b2b8d958871bec9a33953be051a75d8bf07c

Despite thinking this is an improvement on what we currently do and that it's ready for merge, there's two things I'm not 100% comfortable with, and that I think would be good to improve in a followup. (1) I think this puts more demand on outbound peers than it should, as per [previous comments](https://github.com/bitcoin/bitcoin/pull/27626#discussion_r1200426349). (2) I think the multimap stuff makes this hard to understand, and it would be good to have a custom class that abstracts the multimap behaviour. I think it could look something like [this](https://github.com/ajtowns/bitcoin/commit/0d77be6cb2c20466136a99054c5ae696f58f0a55) -- ie, have a structure that collects the results you want from the multimap, and a function that returns that structure that you use in most places. (This also avoids querying the map for the same hash many times in the same function, which isn't actually a problem, but was bugging me...)

mzumsande commented on src/net_processing.cpp:
```diff
@@ -1166,6 +1179,8 @@ bool PeerManagerImpl::BlockRequested(NodeId nodeid, const CBlockIndex& block, st
     CNodeState *state = State(nodeid);
     assert(state != nullptr);
 
+    Assume(mapBlocksInFlight.count(hash) <= MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK);
```
I think that this can currently be broken by `FetchBlock()` / `getblockfrompeer`, which would now allow multiple block requests in parallel, breaking the `MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK` limit. We could probably make sure in `FetchBlock()` that MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK isn't breached (or maybe disallow all parallel request).

instagibbs replied:
restricted it to a single invocation per block at a time. Alternative seems more difficult to think about.

Sjors replied:
It's fine by me, though currently we have no way of telling whether a request is in progress. Maybe bring this up in #27652 too.

instagibbs replied:
After talking offline with @Sjors , his use-case does involve asking multiple peers to detect stale tips.

Instead, any call of `getblockbypeer` will cause all of the same block requests to be forgotten/ignored. It's up to the user of the RPC to not have this interfere with syncing of their own node.

Sjors replied:
That works for me.

theuni reviewed (commented):
Very shallow/mechanical code review ACK. Left a few nits for some head-scratchers.

I was mostly looking for misuses of the multimap. Zero opinion about the logical change itself.

instagibbs commented:
Ready for re-ACKs now that `getblockfrompeer` issue is fixed

sdaftuar commented:
ACK d7f359b35e8b1e9acc4d397de262cd9ba9bbcb83

To reiterate, I think we should continue to keep an eye on bandwidth usage (and load on outbound peers) after this patch and revisit whether any additional tuning of this behavior is needed prior to release. However it'll be useful to have this in master so that we can monitor for the next few months.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -1166,6 +1179,8 @@ bool PeerManagerImpl::BlockRequested(NodeId nodeid, const CBlockIndex& block, st
     assert(state != nullptr);
 
+    Assume(mapBlocksInFlight.count(hash) <= MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK);
+
```
Another nit about the `Assume(mapBlocksInFlight.count(hash) <= MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK);` line - I think that this should this be `<` instead of `<=` because it is the responsibility of the caller to call `BlockRequested` only if there is room for one more request.

instagibbs replied:
caller may call it when it's already requested by this peer, see a few lines below at https://github.com/bitcoin/bitcoin/pull/27626/files/13f9b20b4cb2f3f26e81184a77e9cf1f626d4f57..03423f8bd12b95a06a4a9d8377e781625dd38aae#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3L1175

instagibbs replied:
make sure to check the other call locations, but they aren't supposed to

mzumsande reviewed (commented):
Code Review ACK d7f359b35e8b1e9acc4d397de262cd9ba9bbcb83

I've run this for >2 days now with additional logging and encountered no problems.

One thing I frequently observe that hasn't been mentioned above is that a low-bandwidth peer is first to give us a new header, so there is a header/getdata(CMPCT) roundtrip, and by the time that has finished we already requested the block additionally from a high-bandwidth peer that doesn't need the roundtrip.

fanquake commented:
I also cannot reproduce.

Sjors commented on src/net_processing.cpp:
```diff
@@ -1122,34 +1123,40 @@ bool PeerManagerImpl::IsBlockRequested(const uint256& hash)
 
-    CNodeState *state = State(node_id);
-    assert(state != nullptr);
+        if (from_peer && *from_peer != node_id) {
```
This is the line valgrind trips over. Adding something like `fprintf(stderr, "from_peer=%ld\n", *from_peer);` above it makes it stop complaining. Maybe this is just a valgrind a bug?

Sjors reviewed (commented):
Post merge utACK d7f359b35e8b1e9acc4d397de262cd9ba9bbcb83 modulo question about missing `return`. Other comments can wait for a followup.

Conceptually this PR still relies on our node immediately reacting to an incoming compact block message. This forces us to be a bit impatient and potentially use more bandwidth than needed.

In a followup we could instead track which peers have announced a block and then request block transactions from other peers if the first peer takes too long. This could be a function of the number of transactions we asked for, the typical pace we're used to, if the announcing peer was inbound, etc. That's all much more complex though. See point 4 in #27740.

As @ajtowns mentioned, refactoring the multimap stuff in a followup would really help future developers - including future versions of myself.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -1774,11 +1789,10 @@ std::optional<std::string> PeerManagerImpl::FetchBlock(NodeId peer_id, const CBl
+    // Forget about all prior requests
+    RemoveBlockRequest(block_index.GetBlockHash(), std::nullopt);
+
+    // Mark block as in-flight
```
Not anymore, we completely forget we expect it from a peer now. No disconnect should occur.

Sjors replied:
That was already the case. The comment is there to point out that the current request will result in a disconnect (unless a new request is made on time).

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4361,9 +4383,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+                } else if (pfrom.m_bip152_highbandwidth_to &&
+                    (!pfrom.IsInboundConn() ||
+                    IsBlockRequestedFromOutbound(blockhash) ||
+                    already_in_flight < MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK - 1)) {
```
if it's 1, that's fine, it just won't allow additional slots

Sjors replied:
`> 0` is fine too, but I guess nobody will this to zero accidentally.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4361,9 +4383,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                     fProcessBLOCKTXN = true;
-                } else {
+                } else if (first_in_flight) {
+                    // We will try to round-trip any compact blocks we get on failure,
```
oops, indeed, looks like we'll send off bogus getblocktxns here, if it comes from a preferred peer that's not first. needs a fix.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4343,11 +4360,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+                        return;
+                    } else {
+                        // Give up for this peer and wait for other peer(s)
+                        RemoveBlockRequest(pindex->GetBlockHash(), pfrom.GetId());
```
oops, indeed, looks like we'll send off bogus getblocktxns here, if it comes from a preferred peer that's not first. needs a fix. Wasted bandwidth and should result in "Peer %d sent us block transactions for block we weren't expecting\n"

fanquake commented:
Up for backport in #27752.

pinheadmz commented:
@instagibbs does this close https://github.com/bitcoin/bitcoin/issues/21803 ?
